### ISSUE TITLE
8.11 release

### ISSUE THREAD

**Issue by ppedrot** [coq/coq#10843]
# Release process #

## As soon as the previous version branched off master ##

In principle, these steps should be undertaken by the RM of the next
release. Unfortunately, we have not yet been able to nominate RMs
early enough in the process for this person to be known at that point
in time.

- [X] Create a new issue to track the release process where you can copy-paste
  the present checklist.
- [x] Change the version name to the next major version and the magic
  numbers (see [#7008](https://github.com/coq/coq/pull/7008/files)).

  Additionally, in the same commit, update the compatibility
  infrastructure, which consists of invoking
  [`dev/tools/update-compat.py`](../tools/update-compat.py) with the
  `--master` flag.

  Note that the `update-compat.py` script must be run twice: once
  *immediately after* branching with the `--master` flag (which sets
  up Coq to support four `-compat` flag arguments), *in the same
  commit* as the one that updates `coq_version` in
  [`configure.ml`](../../configure.ml), and once again later on before
  the next branch point with the `--release` flag (see next section).
- [x] Put the corresponding alpha tag using `git tag -s`.
  The `VX.X+alpha` tag marks the first commit to be in `master` and not in the
  branch of the previous version.
- [x] Create the `X.X+beta1` milestone if it did not already exist.
- [x] Decide the release calendar with the team (freeze date, beta date, final
  release date) and put this information in the milestone (using the
  description and due date fields).

## Anytime after the previous version is branched off master ##

- [x] Update the compatibility infrastructure to the next release,
  which consists of invoking
  [`dev/tools/update-compat.py`](../tools/update-compat.py) with the
  `--release` flag; this sets up Coq to support three `-compat` flag
  arguments.  To ensure that CI passes, you will have to decide what
  to do about all test-suite files which mention `-compat U.U` or
  `Coq.Comapt.CoqUU` (which is no longer valid, since we only keep
  compatibility against the two previous versions on releases), and
  you may have to prepare overlays for projects using the
  compatibility flags.

## About one month before the beta ##

- [x] Create the `X.X.0` milestone and set its due date.
- [x] Send an announcement of the upcoming freeze on Coqdev and ask people to
  remove from the beta milestone what they already know won't be ready on time
  (possibly postponing to the `X.X.0` milestone for minor bug fixes,
  infrastructure and documentation updates).
- [x] Determine which issues should / must be fixed before the beta, add them
  to the beta milestone, possibly with a
  ["priority: blocker"](https://github.com/coq/coq/labels/priority%3A%20blocker)
  label. Make sure that all these issues are assigned (and that the assignee
  provides an ETA).
- [x] Ping the development coordinator (**@mattam82**) to get him started on
  the update to the Credits chapter of the reference manual.
  See also [#7058](https://github.com/coq/coq/issues/7058).

  The command that was used in the previous versions to get the list
  of contributors for this version is `git shortlog -s -n
  VX.X+alpha..master | cut -f2 | sort -k 2`. Note that the ordering is
  approximative as it will misplace people with middle names. It is
  also probably not correctly handling `Co-authored-by` info that we
  have been using more lately, so should probably be updated to
  account for this.

## On the date of the feature freeze ##

- [x] Create the new version branch `vX.X` (using this name will ensure that
  the branch will be automatically protected).
- [x] Pin the versions of libraries and plugins in
  `dev/ci/ci-basic-overlays.sh` to use commit hashes or tag (or, if it
  exists, a branch dedicated to compatibility with the corresponding
  Coq branch).
- [x] Remove all remaining unmerged feature PRs from the beta milestone.
- [x] Start a new project to track PR backporting. The project should
  have a "Request X.X+beta1 inclusion" column for the PRs that were
  merged in `master` that are to be considered for backporting, and a
  "Shipped in X.X+beta1" columns to put what was backported. A message
  to **@coqbot** in the milestone description tells it to
  automatically add merged PRs to the "Request ... inclusion" column
  and backported PRs to the "Shipped in ..." column. See previous
  milestones for examples. When moving to the next milestone
  (e.g. X.X.0), you can clear and remove the "Request X.X+beta1
  inclusion" column and create new "Request X.X.0 inclusion" and
  "Shipped in X.X.0" columns.

  The release manager is the person responsible for merging PRs that
  target the version branch and backporting appropriate PRs that are
  merged into `master`.
- [x] Delay non-blocking issues to the appropriate milestone and ensure
  blocking issues are solved. If required to solve some blocking issues,
  it is possible to revert some feature PRs in the version branch only.

## Before the beta release date ##

- [x] Ensure the Credits chapter has been updated.
- [x] Ensure that an appropriate version of the plugins we will distribute with
  Coq has been tagged.
- [x] Have some people test the recently auto-generated Windows and MacOS
  packages.
- [x] In a PR:
  - Change the version name from alpha to beta1 (see
  [#7009](https://github.com/coq/coq/pull/7009/files)).
  - We generally do not update the magic numbers at this point.
  - Set `is_a_released_version` to `true` in `configure.ml`.
- [x] Put the `VX.X+beta1` tag using `git tag -s`.
- [x] Check using `git push --tags --dry-run` that you are not
  pushing anything else than the new tag. If needed, remove spurious
  tags with `git tag -d`. When this is OK, proceed with `git push --tags`.
- [x] Set `is_a_released_version` to `false` in `configure.ml`
  (if you forget about it, you'll be reminded whenever you try to
  backport a PR with a changelog entry).

### These steps are the same for all releases (beta, final, patch-level) ###

- [x] Send an e-mail on Coqdev announcing that the tag has been put so that
  package managers can start preparing package updates (including a
  `coq-bignums` compatible version).
- [x] Ping **@erikmd** to update the Docker images in `coqorg/coq`
  (requires `coq-bignums` in `extra-dev` for a beta / in `released`
  for a final release).
- [x] Draft a release on GitHub.
- [x] Get **@maximedenes** to sign the Windows and MacOS packages and
  upload them on GitHub.
- [x] Prepare a page of news on the website with the link to the GitHub release
  (see [coq/www#63](https://github.com/coq/www/pull/63)).
- [x] Upload the new version of the reference manual to the website.
  *TODO: setup some continuous deployment for this.*
- [x] Merge the website update, publish the release
  and send announcement e-mails.
- [x] Ping **@Zimmi48** to publish a new version on Zenodo.
  *TODO: automate this.*
- [x] Close the milestone

## At the final release time ##

- [x] In a PR:
  - Change the version name from X.X.0 and the magic numbers (see
  [#7271](https://github.com/coq/coq/pull/7271/files)).
  - Set `is_a_released_version` to `true` in `configure.ml`.
- [x] Put the `VX.X.0` tag.
- [x] Check using `git push --tags --dry-run` that you are not
  pushing anything else than the new tag. If needed, remove spurious
  tags with `git tag -d`. When this is OK, proceed with `git push --tags`.
- [x] Set `is_a_released_version` to `false` in `configure.ml`
  (if you forget about it, you'll be reminded whenever you try to
  backport a PR with a changelog entry).

Repeat the generic process documented above for all releases.

- [x] Switch the default version of the reference manual on the website.

## At the patch-level release time ##

We generally do not update the magic numbers at this point (see
[`2881a18`](https://github.com/coq/coq/commit/2881a18)).

**Comment by Blaisorblade** [comment]
Should releasing on opam be added to this checklist? (https://github.com/coq/coq/issues/8446#issuecomment-539613862)

**Comment by ppedrot** [comment]
I don't know whether it's the job of the release manager, but most likely yes.

**Comment by Blaisorblade** [comment]
Maybe it's not (https://gitter.im/coq/coq?at=5d9ccac3e8de6f3ca0610475), but that's a bit strange since opam is documented as supported (and seems the easiest way to install coq-foo libraries from opam). Anyway, I only wish it were part of the release process — far from me to add more work to one person. Cheers!

**Comment by Zimmi48** [comment]
It was decided that it is not really the job of the RM (who may not use opam themselves), but it is the point of this:

>     * Send an e-mail on Coqdev announcing that the tag has been put so that
>       package managers can start preparing package updates (including a
>       `coq-bignums` compatible version).

**Comment by gares** [comment]
May I suggest adding to this checklist to submit opam packages for the beta? https://github.com/ocaml/opam-repository/pull/15017 show that our CI is not catching races in the build system, and fixing the makefile before the final would be better (and less in a hurry).

**Comment by Zimmi48** [comment]
Does the official opam repository accepts beta releases?

**Comment by gares** [comment]
Don't get me started on this... I'd rather use another term to describe their CI.

(do you want to add alpine linux? do you want to uninstall `find` from you linux distro? ...)

**Comment by gares** [comment]
> Does the official opam repository accept beta releases?

I don't know, but I'm pretty sure you can open a PR with title [plz don't merge] just to test if your beta package is OK, and re-push with an updated `src` field in the `opam` file

**Comment by maximedenes** [comment]
> Don't get me started on this... "special"... I'd rather use another term to describe their CI.

But you seemed to say that there are races in our build system that we don't catch. Isn't it worrying, and shouldn't we address it in our own CI, rather than relying on OPAM's CI?

**Comment by gares** [comment]
Oh, that is legitimate. I think their VMs have slow I/O, since I too have `OCAMLC a.ml` followed by `OCAMLOPT a.ml` in my build log, but the two processes don't conflict on `a.cmi`. So in that case artificially slowing down I/O or CPU may work.

What is really unfortunate to me is not that our CI is not good. It is that, in order to release we need to pass CI1 and CI2, but we test CI2 only after tagging. This is easier to solve than having CI1 also cover the cases CI2 covers.

**Comment by gares** [comment]
(also, some of the problems catched by CI2 are just that they test on non standard linux distros, and I don't want that on our CI)

**Comment by Zimmi48** [comment]
FTR the Coq 8.11 project was created a few hours before the branching point, which resulted in two feature PRs being automatically added in the project, despite not needing backporting. I manually removed those. Apart from this, I'm glad that the documentation was enough to allow you to create the backporting project by yourselves :grinning: 

A step was skipped, which is to run the compat update script with the `--released` flag. I'm preparing a PR fixing this.

**Comment by Zimmi48** [comment]
@ppedrot @maximedenes: What is the plan for signing the 8.11.0 Windows and macOS packages? IIUC we are back to Inria IT services signing Windows packages (thus they might not be able to do it during the Christmas break), and we have no singing key for macOS packages. So what are we going to do? Release without packages? Release unsigned packages? Find a way to get signed packages?

(It's truly depressing that this issue isn't fully fixed by now, it has been an issue for as long as I remember participating to this team.)

**Comment by ppedrot** [comment]
AFAIU it's not so much the problem of the RM rather than the poor soul who is supposed to interact with said INRIA IT services, namely @maximedenes.

(BTW the legal and thus official name of the INRIA is still the acronym, giving into their PR bullshit is empowering them.)

**Comment by Zimmi48** [comment]
>AFAIU it's not so much the problem of the RM rather than the poor soul who is supposed to interact with said INRIA IT services

Sure, but as the RM, if you do publish a release announcement, you'll have to note to our poor users that "installers for Windows / macOS will not be available before a few days / weeks (?)".

> (BTW the legal and thus official name of the INRIA is still the acronym, giving into their PR bullshit is empowering them.)

Totally off-topic, but actually the only legal name is the full name "Institut national de recherche en informatique et automatique". The acronym (capitalized or not) has no legal meaning. My work contract says "l'Institut national de recherche en informatique et automatique (...) ci-après désigné « Inria »" and does not contain the capitalized acronym anywhere.

**Comment by ejgallego** [comment]
> installers for Windows / macOS will not be available before a few days / weeks (?)

that should be "signed installers" right? What would be the reason not to include unsigned ones?

**Comment by Zimmi48** [comment]
> > installers for Windows / macOS will not be available before a few days / weeks (?)
> 
> that should be "signed installers" right? What would be the reason not to include unsigned ones?

I don't know. That was one of the options. But if installers are unsigned, this should be marked clearly, for instance by having `-unsigned` in their names. Users might be frightened if they are not aware that they are unsigned and then face the scary Windows or macOS warnings.

**Comment by ejgallego** [comment]
I agree, we should label them unsigned clearly.

**Comment by ppedrot** [comment]
Given that the β release was postponed for one week, I think it is legit to also postpone the 8.11.0 release until January, 6th, if that helps with the INRIA IT team signing the packages and making everybody happy. WDYT?

**Comment by Zimmi48** [comment]
Hum, maybe I should have actually waited on 8.11.2 to be properly announced before closing this :thinking:
When is this going to happen BTW? I still see no macOS installer on the GitHub release.
